src/runtime/ArrayBufferObject.cpp

```diff
@@ -63,72 +63,55 @@ ArrayBufferObject::ArrayBufferObject(ExecutionState& state)
-    , m_context(state.context())
     , m_data(nullptr)
-    , m_bytelength(0)
-    , m_fromExternalMemory(false)
+    , m_byteLength(0)
+    , m_mayPointsSharedBackingStore(false)
 {
-    addFinalizer([](Object* s, void* data) {
-        ArrayBufferObject* self = s->asArrayBufferObject();
-        if (self->m_data && !self->m_fromExternalMemory) {
-            self->m_context->vmInstance()->platform()->onArrayBufferObjectDataBufferFree(self->m_context, self, self->m_data);
-        }
-    },
-                 nullptr);
 }
 
-void ArrayBufferObject::allocateBuffer(ExecutionState& state, size_t bytelength)
+void ArrayBufferObject::allocateBuffer(ExecutionState& state, size_t byteLength)
 {
     detachArrayBuffer();
 
-    ASSERT(bytelength < (size_t)ArrayBufferObject::maxArrayBufferSize);
+    ASSERT(byteLength < (size_t)ArrayBufferObject::maxArrayBufferSize);
 
     const size_t ratio = std::max((size_t)GC_get_free_space_divisor() / 6, (size_t)1);
-    if (bytelength > (GC_get_heap_size() / ratio)) {
+    if (byteLength > (GC_get_heap_size() / ratio)) {
         size_t n = 0;
-        size_t times = bytelength / (GC_get_heap_size() / ratio) / 3;
+        size_t times = byteLength / (GC_get_heap_size() / ratio) / 3;
         do {
             GC_gcollect_and_unmap();
             n += 1;
         } while (n < times);
         GC_invoke_finalizers();
     }
 
-    m_data = (uint8_t*)m_context->vmInstance()->platform()->onArrayBufferObjectDataBufferMalloc(m_context, this, bytelength);
-    m_bytelength = bytelength;
-}
-
-void ArrayBufferObject::attachBuffer(ExecutionState& state, void* buffer, size_t bytelength)
-{
-    detachArrayBuffer();
-
-    // if buffer is null, the ArrayBuffer object still seems detached
-    if (buffer == nullptr) {
-        ASSERT(bytelength == 0);
-        buffer = m_context->vmInstance()->platform()->onArrayBufferObjectDataBufferMalloc(m_context, this, 0);
-    }
-
-    m_data = (uint8_t*)buffer;
-    m_bytelength = bytelength;
+    m_backingStore = new BackingStore(byteLength);
+    m_backingStore->m_isShared = true;
+    m_data = (uint8_t*)m_backingStore->data();
+    m_byteLength = byteLength;
 }
 
-void ArrayBufferObject::attachExternalBuffer(ExecutionState& state, void* buffer, size_t bytelength)
+void ArrayBufferObject::attachBuffer(BackingStore* backingStore)
 {
     detachArrayBuffer();
 
-    m_fromExternalMemory = true;
-    m_data = (uint8_t*)buffer;
-    m_bytelength = bytelength;
+    m_mayPointsSharedBackingStore = true;
+    m_backingStore = backingStore;
+    m_backingStore->m_isShared = true;
+    m_data = (uint8_t*)backingStore->data();
+    m_byteLength = backingStore->byteLength();
 }
 
 void ArrayBufferObject::detachArrayBuffer()
 {
-    if (m_data && !m_fromExternalMemory) {
-        m_context->vmInstance()->platform()->onArrayBufferObjectDataBufferFree(m_context, this, m_data);
+    if (m_data && !m_mayPointsSharedBackingStore) {
+        delete m_backingStore.value();
```

Comment:
m_backingStore is optional, Shouldn't you check if it has a value?

Reply:
If m_data is not null,  the backing store is always not null